src/vga.js

```diff
@@ -740,115 +985,521 @@ VGAScreen.prototype.complete_redraw = function()
+    {
+        return;
+    }
+
+    var horizontal_characters = Math.min(1 + this.horizontal_display_enable_end,
+        this.horizontal_blank_start);
+    var vertical_scans = Math.min(1 + this.vertical_display_enable_end,
+        this.vertical_blank_start);
+
+    if(!horizontal_characters || !vertical_scans)
+    {
+        // Don't update if width or height is zero.
+        // These happen when registers are not fully configured yet.
+        return;
+    }
+
+    if(this.graphical_mode)
+    {
+        this.svga_width = horizontal_characters << 3;
+
+        // Offset is half the number of bytes/words/dwords (depending on clocking mode)
+        // of display memory that each logical line occupies.
+        // However, the number of pixels latched, regardless of addressing mode,
+        // should always 8 pixels per character clock (except for 8 bit PEL width, in which
+        // case 4 pixels).
+        this.virtual_width = this.offset_register << 4;
+
+        // Pixel Width / PEL Width / Clock Select
+        if(this.attribute_mode & 0x40)
+        {
+            this.svga_width >>>= 1;
+            this.virtual_width >>>= 1;
+        }
+
+        this.svga_height = this.scan_line_to_screen_row(vertical_scans);
+
+        // The virtual buffer height is however many rows of data that can fit.
+        // Previously drawn graphics outside of current memory address space can
+        // still be drawn by setting start_address. The address at
+        // VGA_HOST_MEMORY_SPACE_START[memory_space_select] is mapped to the first
+        // byte of the frame buffer. Verified on some hardware.
+        // Depended on by: Windows 98 start screen
+        var available_bytes = VGA_HOST_MEMORY_SPACE_SIZE[0];
+
+        this.virtual_height = Math.ceil(available_bytes / this.vga_bytes_per_line());
+
+        this.set_size_graphical(this.svga_width, this.svga_height, 8,
+            this.virtual_width, this.virtual_height);
+
+        this.update_vertical_retrace();
+        this.update_vga_panning();
+    }
+    else
+    {
+        if(this.max_scan_line & 0x80)
+        {
+            // Double scanning means that half of those scan lines
+            // are just repeats
+            vertical_scans >>>= 1;
+        }
+
+        var height = vertical_scans / (1 + (this.max_scan_line & 0x1F)) | 0;
+
+        if(horizontal_characters && height)
+        {
+            this.set_size_text(horizontal_characters, height);
+        }
     }
+};
 
-    this.bus.send("screen-set-mode", is_graphical);
-    this.stats.is_graphical = is_graphical;
+VGAScreen.prototype.update_vga_panning = function()
+{
+    if(!this.graphical_mode)
+    {
+        this.text_mode_redraw();
+    }
 
-    if(is_graphical)
+    if(this.svga_enabled || !this.virtual_width || !this.svga_width)
     {
-        this.svga_width = width;
-        this.svga_height = height;
-        this.set_size_graphical(width, height, 8);
+        // Skip for svga mode and avoid division by zero
```

Comment:
Should we reinitialise the layers to a single layer here, in case the controller has previously been in vga mode?

Reply:
Good idea. If I'm thinking what you're thinking, that probably means that we should call this `update_vga_panning` function every time the port 1CF dispi write function gets called. That way, when the controller changes between svga <--> vga mode, the layers are setup correctly.

That means that we could rename the `update_vga_panning` function to `update_layers` or something.